File: Data Structures & Algorithms/valid-binary-search-tree/submission-2.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Solution {

    fun isValid(root: TreeNode?, leftMax: Long, rightMax: Long): Boolean {

        if (root == null) {
            return true
        }

        if (root.`val`.toLong() <= leftMax || root.`val`.toLong() >= rightMax) {
            return false
        }

        return isValid(root.left, leftMax, root.`val`.toLong()) && 
        isValid(root.right, root.`val`.toLong(), rightMax)

    }

    fun isValidBST(root: TreeNode?): Boolean {
        return isValid(root, Long.MIN_VALUE, Long.MAX_VALUE)
    }
}

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-2.swift
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     var val: Int
 *     var left: TreeNode?
 *     var right: TreeNode?
 *     init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *     }
 * }
 */

class Solution {
    func goodNodes(_ root: TreeNode?) -> Int {

        func dfs(_ node: TreeNode?, _ maxValue: Int) -> Int {

            if node == nil {
                return 0
            }

            var left = dfs(node?.left, max(maxValue, node!.val))
            var right = dfs(node?.right, max(maxValue, node!.val))

            if node!.val >= maxValue {
                return 1 + left + right
            }
            return left + right

        }

        return dfs(root, Int.min)

    }
}
